prov/efa/src/efa_av.c

```diff
@@ -358,97 +357,98 @@ static int efa_conn_implicit_to_explicit(struct efa_av *av,
 	assert(ofi_genlock_held(&av->util_av.lock));
 	assert(ofi_genlock_held(&av->util_av_implicit.lock));
 
-	util_av_entry =
+	/* Get implicit util AV entry and conn */
+	implicit_util_av_entry =
 	ofi_bufpool_get_ibuf(av->util_av_implicit.av_entry_pool, implicit_fi_addr);
 
-	implicit_av_entry = (struct efa_av_entry *) util_av_entry->data;
+	implicit_av_entry = (struct efa_av_entry *) implicit_util_av_entry->data;
 
 	assert(implicit_av_entry);
 	assert(efa_is_same_addr(
 		raw_addr, (struct efa_ep_addr *) implicit_av_entry->ep_addr));
 
-	conn = &implicit_av_entry->conn;
-	assert(conn->fi_addr == FI_ADDR_NOTAVAIL &&
-	       conn->implicit_fi_addr == implicit_fi_addr);
+	implicit_conn = &implicit_av_entry->conn;
+	assert(implicit_conn->fi_addr == FI_ADDR_NOTAVAIL &&
+	       implicit_conn->implicit_fi_addr == implicit_fi_addr);
 
-	ah = conn->ah;
-	assert(ah);
-	shm_fi_addr = implicit_av_entry->conn.shm_fi_addr;
-
-	efa_av_reverse_av_remove(&av->cur_reverse_av_implicit,
-				 &av->prv_reverse_av_implicit, conn);
+	ah = implicit_conn->ah;
 
-	err = ofi_av_remove_addr(&av->util_av_implicit, implicit_fi_addr);
+	/* Create explicit util AV entry and conn */
+	err = ofi_av_insert_addr(&av->util_av, raw_addr, fi_addr);
 	if (err) {
 		EFA_WARN(FI_LOG_AV,
-			 "ofi_av_remove_addr from implicit AV failed! Error "
+			 "ofi_av_insert_addr into explicit AV failed! Error "
 			 "message: %s\n",
 			 fi_strerror(err));
 		return err;
 	}
 
-	dlist_remove(&implicit_av_entry->conn.implicit_av_lru_entry);
+	explicit_util_av_entry =
+		ofi_bufpool_get_ibuf(av->util_av.av_entry_pool, *fi_addr);
+	explicit_av_entry = (struct efa_av_entry *) explicit_util_av_entry->data;
+	assert(efa_is_same_addr(
+		raw_addr, (struct efa_ep_addr *) explicit_av_entry->ep_addr));
 
-	assert(!dlist_empty(&conn->ah->implicit_conn_list));
-	dlist_remove(&conn->ah_implicit_conn_list_entry);
-	efa_ah_implicit_av_lru_ah_move(av->domain, conn->ah);
+	/* Copy information from implicit conn to explicit conn */
+	explicit_conn = &explicit_av_entry->conn;
+	memset(explicit_conn, 0, sizeof(*explicit_conn));
+	explicit_conn->ep_addr = (struct efa_ep_addr *) explicit_av_entry->ep_addr;
+	assert(av->type == FI_AV_TABLE);
+	explicit_conn->ah = implicit_conn->ah;
+	explicit_conn->fi_addr = *fi_addr;
+	explicit_conn->shm_fi_addr = implicit_conn->shm_fi_addr;
+	explicit_conn->implicit_fi_addr = FI_ADDR_NOTAVAIL;
+	HASH_ITER(hh, implicit_conn->ep_peer_map, map_entry, tmp) {
+		HASH_DELETE(hh, implicit_conn->ep_peer_map, map_entry);
```

Comment:
I forget: how did we update the fi_addr in the peer when the map is in ep? Does it only involve a peer->fi_addr update?